pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy

```diff
@@ -308,21 +308,39 @@ class Utils {
         }
     }
 
+    static void markStageWithTag(String stageName, String stepContextFlowNodeId, String tagName, String tagValue) {
+        CpsThread thread = CpsThread.current()
+        FlowExecution execution = thread.execution
+
+        def stepContextFlowNode = execution.getNode(stepContextFlowNodeId)
+
+        def stageFlowNode = stepContextFlowNode.iterateEnclosingBlocks().find { blockStartNode ->
+            CommonUtils.isStageWithOptionalName(stageName).test(blockStartNode)
+        }
+
+        addTagToFlowNode(stageFlowNode, tagName, tagValue)
+
+    }
+
+    private static void addTagToFlowNode(FlowNode currentNode, String tagName, String tagValue) {
+        if (currentNode != null) {
+            TagsAction tagsAction = currentNode.getAction(TagsAction.class)
+            if (tagsAction == null) {
+                tagsAction = new TagsAction()
+                tagsAction.addTag(tagName, tagValue)
+                currentNode.addAction(tagsAction)
+            } else if (tagsAction.getTagValue(tagName) == null) {
+                tagsAction.addTag(tagName, tagValue)
+                currentNode.save()
+            }
+        }
+    }
+
     static void markStageWithTag(String stageName, String tagName, String tagValue) {
```

Comment:
I've left this in place as it can be called from scripted pipelines (and people definitely do it)